src/nagini_translation/translators/program.py

```diff
@@ -667,6 +674,316 @@ def create_may_set_predicate(self, ctx: Context) -> 'silver.ast.Predicate':
+        declarations.
+        """
+        arguments = []
+        for arg_name, arg_type in cons.fields.items():
+            argument_type = adt_type if arg_type.type.name == adt_name else self.viper.Ref
+            argument = self.viper.LocalVarDecl(arg_name, argument_type, pos, info)
+            arguments.append(argument)
+        return arguments
+
+    def _get_adt_cons_params(self, cons: PythonClass, adt_name: str,
+                             adt_type: DomainType, pos: Position,
+                             info: Info) -> List[Var]:
+        """
+        Returns the parameters of the ADT constructor as a list of variables.
+        """
+        arguments = []
+        for arg_name, arg_type in cons.fields.items():
+            argument_type = adt_type if arg_type.type.name == adt_name else self.viper.Ref
+            argument = self.viper.LocalVar(arg_name, argument_type, pos, info)
+            arguments.append(argument)
+        return arguments
+
+    def _conjoin(self, eqs: List[Expr], pos: Position, info: Info) -> Expr:
+        """
+        Conjoin all expressions in the list.
+        """
+        return eqs[0] if len(eqs) == 1 else self.viper.And(eqs[0],
+               self._conjoin(eqs[1:], pos, info), pos, info)
+        
+    def _disjoin(self, eqs: List[Expr], pos: Position, info: Info) -> Expr:
+        """
+        Disjoin all expressions in the list.
+        """
+        return eqs[0] if len(eqs) == 1 else self.viper.Or(eqs[0],
+               self._disjoin(eqs[1:], pos, info), pos, info)
+
+    def _create_adt_func_constructors(self, adt: PythonClass, adt_prefix: str,
+                                      adt_name: str, adt_type: DomainType,
+                                      pos: Position, info: Info) -> List[DomainFunc]:
+        """
+        Create domain functions representing constructors of the ADT.
+        """
+        assert adt.all_subclasses[0] == adt
+        domain_funcs = []
+        for cons in adt.all_subclasses[1:]:
+            arguments = self._get_adt_cons_param_decl(cons, adt.name, adt_type, pos, info)
+            yield self.viper.DomainFunc(adt_prefix + cons.name, arguments,
+                                        adt_type, False, pos, info, adt_name)
+    
+    def _create_adt_func_deconstructors(self, adt: PythonClass, adt_prefix: str,
+                                        adt_name: str, adt_type: DomainType,
+                                        pos: Position, info: Info) -> List[DomainFunc]:
+        """
+        Create domain functions representing deconstructors of the ADT.
+        """
+        adt_obj_decl = self.viper.LocalVarDecl('obj', adt_type, pos, info)
+        for cons in adt.all_subclasses[1:]:
+            for arg_name, arg_type in cons.fields.items():
+                function_type = adt_type if arg_type.type.name == adt.name else self.viper.Ref
+                yield self.viper.DomainFunc(adt_prefix + cons.name + '_' + arg_name,
```

Comment:
I think I also wrote this somewhere else that you haven't responded to yet, but it looks like the names here aren't generated using the fresh names functionality, and they probably should be.